### PR TITLE
feat: add location & iam mixin services

### PR DESCRIPTION
**Pull request by noahdietz**
Adds Location and IAMPolicy mix-in service gRPC implementations.

Adds an API Service configuration file with mix-in APIs and their `http.rules`.

### CODE REVIEW

**Comment by noahdietz** [comment]
@vchudnov-g I'm wondering what you think we should do for the Rest handlers. Shall I work to add generation of the handlers for the mix-ins to `genrest`? I need to add a `showcase_v1beta1.yaml` service config anyways, and we could have it consume that for the mix-in `http.rules`?

**Comment by noahdietz** [comment]
> Shall I work to add generation of the handlers for the mix-ins to `genrest`...we could have it consume [service config] for the mix-in http.rules

I've just added a Service config file. Since it is full of `http.rules` it should be easy to add support for them in `genrest`. WDYT?

We could also add in another PR.

File: server/services/iam_policy_service.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+	"github.com/golang/protobuf/proto"
+	iampb "google.golang.org/genproto/googleapis/iam/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var missingResource error = status.Error(codes.InvalidArgument, "Missing required argument: resource")
+var missingPolicy error = status.Error(codes.InvalidArgument, "Missing required argument: policy")
+var resourceDNE error = status.Error(codes.NotFound, "Requested resource has no Policy")
```
Comment (vchudnov-g) [inline server/services/iam_policy_service.go:29]: Is DNE a standard acronym? ("do not erase"?)

File: server/services/iam_policy_service.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	defer i.mu.Unlock()
+
+	clone := proto.Clone(in.GetPolicy())
+	p := clone.(*iampb.Policy)
+	i.policies[in.GetResource()] = p
+
+	return in.GetPolicy(), nil
+}
+
+// TestIamPermissions verifies that the requested resource has been Set at one point, and echos the back permissions to be tested.
```
Comment (vchudnov-g) [inline server/services/iam_policy_service.go:80]: nit: "echoes" here and elsewhere in this PR

File: server/services/iam_policy_service.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+// NewIAMPolicyServer returns a new LocationsServer for the Showcase API.
+func NewIAMPolicyServer() iampb.IAMPolicyServer {
+	return &iamPolicyServerImpl{
+		policies: map[string]*iampb.Policy{},
+	}
+}
+
+type iamPolicyServerImpl struct {
+	mu       sync.Mutex
+	policies map[string]*iampb.Policy
```
Comment (vchudnov-g) [inline server/services/iam_policy_service.go:40]: nit: comment that the key is the resource

File: server/services/iam_policy_service_test.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+			resource: "foo/bar",
+			err:      resourceDNE,
+		},
+	} {
+		res, err := s.GetIamPolicy(context.Background(), &iampb.GetIamPolicyRequest{Resource: tst.resource})
+		if err != nil {
+			if diff := cmp.Diff(err, tst.err, cmpopts.EquateErrors()); diff != "" {
+				t.Errorf("GetIamPolicy_errors(%s): got(-),want(+):\n%s", tst.name, diff)
+			}
+		} else {
```
Comment (vchudnov-g) [inline server/services/iam_policy_service_test.go:92]: nit: per idiomaticity of getting errors out of the way and minimizing conditional branches and nesting, consider flipping this and the other negative tests (in this file and the others):

```
if err == nil {
  ...
  continue
}
if diff := ... {
  ...
}
```

File: server/services/locations_service.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+// NewLocationsServer returns a new LocationsServer for the Showcase API.
+func NewLocationsServer() locpb.LocationsServer {
+	return &locationsServerImpl{}
+}
+
+type locationsServerImpl struct{}
+
+// GetLocation echos back a Location with the same name as
+// GetLocationRequest.name and has the last segment of name as the display_name.
```
Comment (vchudnov-g) [inline server/services/locations_service.go:36]: nit: Should we say ` as in.name` instead? Similarly below.

File: server/services/locations_service_test.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	locpb "google.golang.org/genproto/googleapis/cloud/location"
+	"google.golang.org/protobuf/proto"
+)
+
+func TestListLocations(t *testing.T) {
+	s := NewLocationsServer()
+	foo := "projects/foo"
```
Comment (vchudnov-g) [inline server/services/locations_service_test.go:29]: nit: why not s/`foo`/`name`/ or s/`foo`/`locName`/ (for the variable name)?

**Review by vchudnov-g — APPROVED** [review]
Thanks for doing this! Nothing blocking, only minor comments and nits.

File: schema/google/showcase/v1beta1/showcase_v1beta1.yaml [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+title: The Client Libraries Showcase API
+
+apis:
+- name: google.showcase.v1beta1.Compliance
+- name: google.showcase.v1beta1.Echo
+- name: google.showcase.v1beta1.Identity
+- name: google.showcase.v1beta1.Messaging
+- name: google.showcase.v1beta1.SequenceService
+- name: google.showcase.v1beta1.Testing
+# Mix-in services
```
[the remark this answers is not in the document]
Comment (noahdietz) [reply schema/google/showcase/v1beta1/showcase_v1beta1.yaml:13]: I'm happy to put a license in it. We don't have a license header in any of our other published service yaml files. Do you think that should change? Might be worth asking the team...

[the remark this answers is not in the document]
Comment (vchudnov-g) [reply schema/google/showcase/v1beta1/showcase_v1beta1.yaml:13]: I tend to err on the side of caution: if a format allows comments, then I try to put the necessary license in (disclaimer: I'm probably not consistent). We can discuss with the team. I would imagine that shouldn't block this PR, given precedent...but IANAL.

[the remark this answers is not in the document]
Comment (noahdietz) [reply schema/google/showcase/v1beta1/showcase_v1beta1.yaml:13]: Better to be safe than sorry, happy to add one!

File: server/services/iam_policy_service.go (hunk 1 above)
Comment (noahdietz) [reply server/services/iam_policy_service.go:29]: Yeah, "does not exist". But I will write it out, acronyms aren't inclusive.

File: server/services/iam_policy_service.go (hunk 2 above)
Comment (noahdietz) [reply server/services/iam_policy_service.go:80]: Done.

File: server/services/iam_policy_service.go (hunk 3 above)
Comment (noahdietz) [reply server/services/iam_policy_service.go:40]: Done.

File: server/services/iam_policy_service_test.go (hunk 4 above)
Comment (noahdietz) [reply server/services/iam_policy_service_test.go:92]: Good call, thank you. Done!

File: server/services/locations_service.go (hunk 5 above)
Comment (noahdietz) [reply server/services/locations_service.go:36]: Done.

File: server/services/locations_service_test.go (hunk 6 above)
Comment (noahdietz) [reply server/services/locations_service_test.go:29]: I went with `name`. Done!

**Pull request merged**